Source/JavaScriptCore/runtime/StringPrototype.cpp

```diff
@@ -576,57 +570,147 @@ static ALWAYS_INLINE JSString* replaceUsingRegExpSearchWithCache(VM& vm, JSGloba
-            JSString* matchedString = asString(result->get(cursor + 0));
-            JSValue startOffset = result->get(cursor + 1);
+        uint64_t totalLength = 0;
+        bool replacementsAre8Bit = true;
+        {
+            size_t lastIndex = 0;
+            size_t index = 0;
+            for (auto& slot : replacements) {
+                JSString* matchedString = asString(result->get(index * 2));
+                JSValue startOffset = result->get(index * 2 + 1);
+
+                int32_t start = startOffset.asInt32();
+                int32_t end = start + matchedString->length();
+
+                JSValue jsResult = cachedCall.callWithArguments(globalObject, jsUndefined(), matchedString, startOffset, string);
+                RETURN_IF_EXCEPTION_WITH_TRAPS_DEFERRED(scope, nullptr);
+
+                auto string = jsResult.toWTFString(globalObject);
+                RETURN_IF_EXCEPTION_WITH_TRAPS_DEFERRED(scope, nullptr);
+
+                replacementsAre8Bit &= string.is8Bit();
+                totalLength += string.length();
+                totalLength += (start - lastIndex);
+                slot = WTFMove(string);
+
+                lastIndex = end;
+                ++index;
+            }
+            if (static_cast<unsigned>(lastIndex) < sourceLen)
+                totalLength += (sourceLen - lastIndex);
+        }
 
-            int32_t start = startOffset.asInt32();
-            int32_t end = start + matchedString->length();
+        if (UNLIKELY(totalLength > StringImpl::MaxLength)) {
+            throwOutOfMemoryError(globalObject, scope);
+            return nullptr;
+        }
 
-            JSValue jsResult = cachedCall.callWithArguments(globalObject, jsUndefined(), matchedString, startOffset, string);
-            RETURN_IF_EXCEPTION_WITH_TRAPS_DEFERRED(scope, nullptr);
+        StringView sourceView { source };
+        if (sourceView.is8Bit() && replacementsAre8Bit) {
+            std::span<LChar> buffer;
+            auto impl = StringImpl::tryCreateUninitialized(totalLength, buffer);
+            if (UNLIKELY(!impl)) {
+                throwOutOfMemoryError(globalObject, scope);
+                return nullptr;
+            }
 
-            auto string = jsResult.toWTFString(globalObject);
-            RETURN_IF_EXCEPTION_WITH_TRAPS_DEFERRED(scope, nullptr);
+            size_t lastIndex = 0;
+            unsigned index = 0;
+            size_t bufferPos = 0;
+            for (auto& replacement : replacements) {
+                int32_t length = asString(result->get(index * 2))->length();
+                int32_t start = result->get(index * 2 + 1).asInt32();
+                int32_t end = start + length;
 
-            sourceRanges.constructAndAppend(lastIndex, start);
-            replacements[replacementIndex++] = WTFMove(string);
+                auto substring = sourceView.substring(lastIndex, start - lastIndex);
+                substring.getCharacters8(buffer.subspan(bufferPos));
```

Comment:
lol, `getCharacters8` the naming is confusing. Maybe `copyCharacters8` is better.

Reply:
Anyway, this is pre-existing code :)